File: VB/Controllers/HomeController.vb
Imports System.Collections
Imports System.Web.Mvc

Namespace GridViewChangeModelMvc.Controllers
	Public Class HomeController
		Inherits Controller

		Public Function Index() As ActionResult
			Return View()
		End Function

		Public Function GridViewPartial(Optional ByVal gridType As GridType = GridType.Categories) As ActionResult
			Dim model As IEnumerable = Nothing

			Select Case gridType
				Case GridType.Categories
					model = NorthwindDataProvider.GetCategories()
					ViewData("KeyFieldName") = "CategoryID"
				Case GridType.Products
					model = NorthwindDataProvider.GetProducts()
					ViewData("KeyFieldName") = "ProductID"
			End Select

			Return PartialView(model)
		End Function
	End Class
End Namespace

File: CS/Controllers/HomeController.cs
using System.Collections;
using System.Web.Mvc;

namespace GridViewChangeModelMvc.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();
        }

        public ActionResult GridViewPartial(GridType gridType = GridType.Categories) {
            IEnumerable model = null;

            switch (gridType) {
                case GridType.Categories:
                    model = NorthwindDataProvider.GetCategories();
                    ViewData["KeyFieldName"] = "CategoryID";
                    break;
                case GridType.Products:
                    model = NorthwindDataProvider.GetProducts();
                    ViewData["KeyFieldName"] = "ProductID";
                    break;
            }

            return PartialView(model);
        }
    }
}